jdtls.ext/com.microsoft.jdtls.ext.core/src/org/eclipse/jdt/internal/jarpackager/JarPackageUtil.java

```diff
@@ -0,0 +1,356 @@
+package org.eclipse.jdt.internal.jarpackager;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+import java.util.zip.CRC32;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+import java.util.zip.ZipFile;
+
+import org.eclipse.core.runtime.*;
+import org.eclipse.jdt.internal.core.manipulation.JavaManipulationPlugin;
+import org.eclipse.jdt.internal.ui.IJavaStatusConstants;
+
+public class JarPackageUtil {
+
+	/**
+	 * Write the given entry describing the given content to the current archive.
+	 * Extracted from org.eclipse.jdt.ui.jarpackager.JarWriter3
+	 *
+	 * @param   entry            the entry to write
+	 * @param   content          the content to write
+	 * @param   fJarOutputStream the destination JarOutputStream
+	 *
+	 * @throws IOException       If an I/O error occurred
+	 *
+	 * @since 1.14
+	 *
+	 */
+	public static void addEntry(JarEntry entry, InputStream content, JarOutputStream fJarOutputStream) throws IOException {
+		byte[] readBuffer= new byte[4096];
+		try {
+			fJarOutputStream.putNextEntry(entry);
+			int count;
+			while ((count= content.read(readBuffer, 0, readBuffer.length)) != -1)
+				fJarOutputStream.write(readBuffer, 0, count);
+		} finally  {
+			if (content != null)
+				content.close();
+
+			/*
+			 * Commented out because some JREs throw an NPE if a stream
+			 * is closed twice. This works because
+			 * a) putNextEntry closes the previous entry
+			 * b) closing the stream closes the last entry
+			 */
+		}
+	}
+
+	/**
+	 * Write the contents of the given zipfile to the JarOutputStream.
+	 * Extracted from org.eclipse.jdt.internal.ui.jarpackagerfat.UnpackFatJarBuilder
+	 *
+	 * @param   zipFile          			the zipfile to extract
+	 * @param   areDirectoryEntriesIncluded Tells whether directory entries are added to the jar
+	 * @param   isCompressed				whether the jar is compressed or not
+	 * @param   fJarOutputStream            the destination JarOutputStream
+	 * @param	fDirectories				the temporary set saves existing directories
+	 * @param   progressMonitor  			the progressMonitor
+	 *
+	 * @return                   			the MultiStatus saving the warnings during the process
+	 *
+	 * @since 1.14
+	 *
+	 */
+	public static MultiStatus writeArchive(ZipFile zipFile, boolean areDirectoryEntriesIncluded,
+								  boolean isCompressed, JarOutputStream fJarOutputStream,
+								  Set<String> fDirectories, IProgressMonitor progressMonitor) {
+		MultiStatus fStatus =  new MultiStatus(JavaManipulationPlugin.getPluginId(), IStatus.OK, ""); //$NON-NLS-1$
```

Comment:
Please use JdtlsExtActivator.PLUGIN_ID instead of JavaManipulationPlugin.getPluginId()